### ISSUE TITLE
How to actually use the certs?

### ISSUE THREAD

**Issue by j1elo** [FiloSottile/mkcert#60]
Hi, totally newbie question here. I'm not a web developer and this is the first time I configure self-signed certificates, so bear with me. Also I'd like to propose adding a section in the documentation for people in my situation.

This is what I'm currently doing, step by step command-line style. Server is an Amazon AWS machine with Ubuntu 16.04, in which I'm doing some WebRTC tests; Chrome and Firefox will refuse to allow webcam and microphone access to insecure sites (except for `localhost`), so I need to serve an HTTPS page from my test server:

```
# [On SERVER] Download and prepare 'mkcert':
mkdir ~/mkcert
cd ~/mkcert
wget https://github.com/FiloSottile/mkcert/releases/download/v1.1.0/mkcert-v1.1.0-linux-amd64
mv mkcert-v*-linux-amd64 mkcert
chmod +x mkcert

# [On SERVER] Create a CA used for signing certificates:
sudo apt-get install -y libnss3-tools
./mkcert -install
cp "$(./mkcert -CAROOT)/rootCA.pem" ./

# [On SERVER] Create certificate for needed domains:
./mkcert '*.compute.amazonaws.com' localhost 127.0.0.1

# [On SERVER] Start HTTPS server using Node.js:
curl -sL https://deb.nodesource.com/setup_8.x | sudo -E bash -
sudo apt-get install -y nodejs
sudo npm install -g http-server
http-server --ssl \
    --cert ~/mkcert/_wildcard.compute.amazonaws.com+2.pem \
    --key  ~/mkcert/_wildcard.compute.amazonaws.com+2-key.pem

# [On CLIENT] Copy required files:
sudo apt-get install -y libnss3-tools
rsync -av ${SERVER_IP}:mkcert ~/
cd ~/mkcert
CAROOT="$PWD" ./mkcert -install
```

At this point, I open this URL in Chrome:
`https://MY-SERVER-NAME.compute.amazonaws.com:8080/

But it still shows a warning page right before loading, and after dismissing the warning, a RED warning with "Not secure" text is shown in the address bar.

What I expected is that Chrome loads the page without any security warnings and with a GREEN lock in the address bar.

What steps I'm missing to make this work as intended?

**Comment by j1elo** [comment]
More info:

Chrome says this in its warning screen:
```
NET::ERR_CERT_COMMON_NAME_INVALID
This server could not prove that it is MY-SERVER-NAME.compute.amazonaws.com; its security certificate is from *.compute.amazonaws.com.
```

And Firefox says something similar:
```
MY-SERVER-NAME.compute.amazonaws.com uses an invalid security certificate.
The certificate is only valid for the following names: *.compute.amazonaws.com, localhost, 127.0.0.1
Error code: SSL_ERROR_BAD_CERT_DOMAIN
```

Does this mean that it's not possible to create certificates for `AnySubdomain.domain.com`?

I wanted to do this because the name that AWS gives your machine (what I call `MY-SERVER-NAME` in my code) changes every time the machine starts up, so the best would be to have a certificate that doesn't mind what is the actual name of the subdomain.

**Comment by nickkaczmarek** [comment]
I may be misunderstanding this, but since you're hitting your site from the `amazonaws.com` tld, mkcert won't work for this. You'll probably need to use let's encrypt or something and put that certificate on your aws server. Someone who knows more may have more insight, but that's how I understand this. This might be helpful, https://docs.aws.amazon.com/acm/latest/userguide/setup-website.html

**Comment by neoKushan** [comment]
I don't believe you're using the tool as it was intended, as it's very much aimed at generating certs for *local* development. For deploying to AWS (or any server, really) you are best looking into something like Let's Encrypt for genuine SSL (or using the SSL cert that the cloud provider usually gives you). I don't know enough about aws to help with this, but there's plenty of documentation out there.

**Comment by j1elo** [comment]
I see no reason why these certs wouldn't work for external machines; even the [README](https://github.com/FiloSottile/mkcert/blob/53f1769ab57d2e272477af28d19113e121718839/README.md) of this project starts by showing how to generate a cert for such one! (example.com)
`$ mkcert example.com '*.example.org' myapp.dev localhost 127.0.0.1 ::1`

I'm not a web dev and had zero idea of how to configure a certificate, so I just wanted to have a pair of files that can be copied to whatever machine, regardless of it being local or remote, and instantly have a valid HTTPS connection to them... well, that was the objective. Maybe I should have looked into Let's Encrypt for the AWS machine.

In any case, I actually made it work without any further problem whatsoever. The solution to my problem was that *a restriction exists on how the wildcard certificates work by spec*, not anything to do specifically with mkcert.

It turns out that a wildcard such `*.example.com` **won't match sub-subdomains** such as `a.b.example.com`. As simple as that.

So I used `mkcert` to generate a certificate for `*.region.compute.amazonaws.com`. Problem solved.

I won't close this issue yet to allow the author see it and consider my proposal of adding a section in the documentation that talks about this use case. @FiloSottile thank you for this tool!

**Comment by Suleman-Elahi** [comment]
Doesn't help

![image](https://user-images.githubusercontent.com/24384996/44509095-3061f500-a6ce-11e8-8213-93c550fd9233.png)

**Comment by j1elo** [comment]
That error doesn't depend on running with or without Administrator CMD. It happens because [a regular expression fails to match](https://github.com/FiloSottile/mkcert/blob/060fcce2db48466c3903803e3a42b2689ace1617/main.go#L138).

Note how the error says: `ERROR: "'*.example.org'"` it is including the single quotes inside the double quotes. Of course, `'*.example.org'` is an invalid hostname. Try `*.example.org` without quotes. Or with double quotes. It all depends on how the CMD interpreter treats simple (and double) quotes in call arguments.

That's probably a documentation bug in mkcert. Please handle it in a new issue, and don't derail already existing ones such as this one.

**Comment by Suleman-Elahi** [comment]
Thank you very much !!!! @j1elo
